### PR TITLE
s4702040 Improved UNet - ISIC 2018 Task 1 Segmentation

### PR DESCRIPTION
**Pull request by JRSantos4742**
Name: Joshua Santos
ID: s4702040

Hi, I am submitting my implementation of the Improved UNet for the ISIC 2018 segmentation dataset.

The result was a dice coefficient of 0.79 which was below the target value of 0.8.

The files to be added are:

- dataset.py which defines custom dataset class for loading image and mask data
- module.py which defines the pytorch implementation of the improved unet
- predict.py which predicts a segmentation mask for a single image using a trained model
- train.py which trains and evaluates model as well as saves/loads the model and plots the losses
- several images for readme such as loss plot and predicted masks

### CODE REVIEW

**Comment by wangzhaomxy** [comment]
This is an initial inspection, no action is required at this point.

- Well organized files.
- Algorithm solves the problem appropriately. Average Dice Coefficient in testing dataset: 0.78. Predicted Mask output looks good. No data augmentation. Other parts are ok.
- Using Pytorch to construct the Improved U-Net models and functions. However, the model body is roughly an improved U-Net model(filters start with 8 and end up with 128) instead of the improved U-Net in the paper(filters start with 16 and end up with 256). Other parts are ok.
- Good coding design. 
- Sufficient docstring and comment in the code, good references. 
- Easy difficulty problem, no higher difficulty marks.

**Comment by JRSantos4742** [comment]
> * Well organized files.
> * Algorithm solves the problem appropriately. Average Dice Coefficient in testing dataset: 0.78. Predicted Mask output looks good. No data augmentation. Other parts are ok.
> * Using Pytorch to construct the Improved U-Net models and functions. However, the model body is roughly an improved U-Net model(filters start with 8 and end up with 128) instead of the improved U-Net in the paper(filters start with 16 and end up with 256). Other parts are ok.
> * Good coding design.
> * Sufficient docstring and comment in the code, good references.
> * Easy difficulty problem, no higher difficulty marks.

Should I be changing the filters for the model or is it fine as is?

**Comment by wangzhaomxy** [comment]
> > * Well organized files.
> > * Algorithm solves the problem appropriately. Average Dice Coefficient in testing dataset: 0.78. Predicted Mask output looks good. No data augmentation. Other parts are ok.
> > * Using Pytorch to construct the Improved U-Net models and functions. However, the model body is roughly an improved U-Net model(filters start with 8 and end up with 128) instead of the improved U-Net in the paper(filters start with 16 and end up with 256). Other parts are ok.
> > * Good coding design.
> > * Sufficient docstring and comment in the code, good references.
> > * Easy difficulty problem, no higher difficulty marks.
> 
> Should I be changing the filters for the model or is it fine as is?

It's fine. My comment is just for marking.

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage, could have used augmentation -1
Difficulty: Easy -10
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable, no usage info -1
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
